`storage`: more cached disk usage improvements

WillemKauf opened the pull request:
Quick follow-up to https://github.com/redpanda-data/redpanda/pull/26513. Only last 6 commits are new. Other PR should be reviewed first.

### Commit ac16830ea96697f2477c3a9677188ffba7289fc1

Sets offsets in `segment_appender::close()`.

We were not properly setting the book-kept offsets in the `segment_appender` while closing it. This commit fixes that so `_committed_offset`, `_fallocation_offset`, `_flushed_offset`, and `_stable_offset` will all report the same value after a `segment_appender` is closed (previously, the latter 3 would report untruncated/unflushed state). This will be relevant in the next commit.

### Commit 11129fe3f4238a59c68aedbfbb791a9b47e66e37

Adds `segment_appender::size_bytes()`.

Previously, uses of `_appender->file_byte_offset()` were used to set the cached file size after a `segment_appender` was closed. However, while a `segment_appender` is open, this function returns the position of the cursor within the file, not the file size itself (as we `fallocate()` up front in the `segment_appender`). To properly report the size of a file with an open `segment_appender`, we should be returning `_fallocation_offset` instead. With the fix from the previous commit, this offset will also now be accurate after the `segment_appender` is closed.

### Commit 23ff40fd2dffbf5a12fb034c9f93c2074c6c847c

Use `segment_appender::size_bytes()` in `spill_key_index::size_bytes()` instead of `segment_appender::file_byte_offset()` for the aforementioned reason of accurate file size reporting while the `appender` is open.

### Commit db02ffd9780692e3fb775f36ace93cf63e6cc891

This commit seeks to cut down on the number of syscalls made due to calls to `persistent_size()` on an open `segment`. We clear the cached values for open `segment`s during a lot of operations- 
* `do_flush()`
* `truncate()`
* `do_append()`
* `advance_stable_offset()`

to name a few. While a `segment` is still open, we have access to its `appender` and its `compacted_index_writer`, which means we can simply defer to these objects' `size_bytes()` [aka `_fallocation_offset`] instead of issuing a syscall.

### Commit 0c22b4a2428ef1bed434dc481e8883e461124ff5

Just a simple comment tweak for `set_cached_disk_usage()`.

### Commit 1873fba2e38f9e02f7bcf13f14f0c0b479ef1476

An optimization in `do_release_appender()`, which has the effect of leaving a freshly closed `segment` with cached disk usages intact instead of cleared (since this is the last time we will have access to its `appender` and `compacted_index_writer`).

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

### Improvements

* Cut down the amount of time spent in `fstat()` syscalls in the `storage` layer EVEN MORE IN GENERAL!

WillemKauf committed:
`storage`: use appender sizes if possible in `segment::persistent_size()`

Before a `segment` is closed and `appender`s are still owned, they can be
used to provide an estimate of the underlying file size on disk without
needing to make a call to `ss::file_size()`.

Use these in `persistent_size()` to cut down on the number of syscalls
to `ss::file_size()` while a `segment` is still open (and constantly
invalidating its cached sizes through various `segment` operations).

WillemKauf committed:
`storage`: edit `set_cached_disk_usage()` comment

To better describe the current use and optional setting
using the `new_compacted_index_size`.

WillemKauf committed:
`storage`: optimize `do_release_appender()` for cached disk usage

There are two major changes here:
1. Move the call to `clear_cached_disk_usage()` above the call to
   `_idx.flush()`, as the latter will set the index's `cached_disk_usage`
   value during its operation, and clearing it immediately afterwards is
   sub-optimal (as it guarantees an `ss::file_size()` call will be needed
   in the future).
2. Add a call to `set_cached_disk_usage()` after both `appender` and
   `compacted_index` are `close()`'d.

These changes will ensure that the closed `segment` will have all of its
cached disk usage values set, and a future call to `persistent_size()` won't
need to go through a syscall.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/67637-->
<!--b23ffa6d-f255-4f7b-916c-032da40048d3-->
<!--df24cd52-5dad-4815-8d39-c5f7c349c992-->
<!--b66fe44a-b2a2-4521-a6b6-fbb563ad6ec2-->
<!--9af9c07d-0c25-4d7e-98a9-2800ba593a2b-->
<!--title-->
## Retry command for [Build#67637](https://buildkite.com/redpanda/redpanda/builds/67637)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/full_disk_test.py::LocalDiskReportTest.test_basic_usage_report
tests/rptest/tests/datalake/blocked_catalog_test.py::DatalakeBlockedCatalogTest.test_trim_local_after_first_translation@{"cloud_storage_type":1}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--d3eec222-5d66-4cab-923d-3cb5e9a8973d-->
<!--TEST RESULTS-->
<!--60527387-c2aa-4a73-b0ae-3ec74eab060b-->
<!--TEST RESULTS-->
<!--ce43085e-8508-45d2-87cc-c903b8939a9a-->
<!--TEST RESULTS-->
<!--4b721548-5f60-4ca6-acdb-de1ed4a19cac-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67637">build#67637</a></summary>


| test_class                    | test_method                               | test_arguments                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:------------------------------|:------------------------------------------|:-------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ConsumerOffsetsRecoveryTest   | test_consumer_offsets_partition_recovery  |                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67637#01978b64-d3ed-430a-b9c0-764327b93f54 | FLAKY         | 19/21    | upstream reliability is '97.96437659033079'. current run reliability is '90.47619047619048'. drift is 7.48819 and the allowed drift is set to 50. The test should PASS  |
| DatalakeBlockedCatalogTest    | test_trim_local_after_first_translation   | {"cloud_storage_type": 1}                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67637#01978b64-d3ed-430a-b9c0-764327b93f54 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
| DatalakeBlockedCatalogTest    | test_trim_local_after_first_translation   | {"cloud_storage_type": 1}                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67637#01978b67-80e3-41df-8715-dcc8d57b9b33 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
| DatalakeDelayedEnablementTest | test_enabling_iceberg_in_existing_cluster | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67637#01978b67-80e3-4ea1-ae59-d6469bf80e21 | FLAKY         | 14/21    | upstream reliability is '78.09330628803245'. current run reliability is '66.66666666666666'. drift is 11.42664 and the allowed drift is set to 50. The test should PASS |
| LocalDiskReportTest           | test_basic_usage_report                   |                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67637#01978b64-d3ee-4a08-a737-cd602ca660d5 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
| LocalDiskReportTest           | test_basic_usage_report                   |                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67637#01978b67-80e2-44e7-a9af-def47720ce48 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  |
| MaintenanceTest               | test_maintenance_sticky                   | {"use_rpk": false}                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67637#01978b67-80e3-4ea1-ae59-d6469bf80e21 | FLAKY         | 20/21    | upstream reliability is '96.88385269121812'. current run reliability is '95.23809523809523'. drift is 1.64576 and the allowed drift is set to 50. The test should PASS  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67671">build#67671</a></summary>


| test_class              | test_method                             | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:------------------------|:----------------------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ArchivalTest            | test_all_partitions_leadership_transfer | {"cloud_storage_type": 1} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67671#01978ed1-412b-4627-a26a-b77dd78a7e55 | FLAKY         | 20/21    | upstream reliability is '99.63369963369964'. current run reliability is '95.23809523809523'. drift is 4.3956 and the allowed drift is set to 50. The test should PASS  |
| DisablingPartitionsTest | test_disable                            |                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67671#01978ed0-4135-44f3-a2f4-f8bac2a802a6 | FLAKY         | 19/21    | upstream reliability is '93.92712550607287'. current run reliability is '90.47619047619048'. drift is 3.45094 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67838">build#67838</a></summary>


| test_class               | test_method                        | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:-----------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ArchivalTest             | test_connection_flicker            | {"cloud_storage_type": 1}                                                                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67838#0197a57a-edff-4888-9509-e3012fb4e153 | FLAKY         | 20/21    | upstream reliability is '99.6551724137931'. current run reliability is '95.23809523809523'. drift is 4.41708 and the allowed drift is set to 50. The test should PASS  |
| DataMigrationsApiTest    | test_migrated_topic_data_integrity | {"params": {"cancellation": null, "use_alias": true}, "transfer_leadership": true}                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67838#0197a57a-edfd-4e6e-9d58-c40db84eaa72 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| RandomNodeOperationsTest | test_node_operations               | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": false, "with_iceberg": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67838#0197a57e-c1b3-49bf-b04a-42789ece622a | FLAKY         | 20/21    | upstream reliability is '99.37106918238993'. current run reliability is '95.23809523809523'. drift is 4.13297 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67858">build#67858</a></summary>


| test_class      | test_method             | test_arguments    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------|:------------------------|:------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MaintenanceTest | test_maintenance_sticky | {"use_rpk": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67858#0197a6c3-a786-4a9d-8e0f-de222597da7f | FLAKY         | 20/21    | upstream reliability is '98.23232323232324'. current run reliability is '95.23809523809523'. drift is 2.99423 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

WillemKauf committed:
`storage`: set `offset`s in `segment_appender::close()`

Here, we perform a flush & truncation of the underlying file, but were not
updating the corresponding offsets in the `segment_appender`.

This means that a closed `appender` will be reporting the untruncated/flushed
offsets if queried.

Fix the book-keeping by setting the `_fallocation_offset`, `_flushed_offset` and
`_stable_offset` after truncating in `close()`.

WillemKauf committed:
`storage`: add `segment_appender::size_bytes()`

This function seeks to report the actual size of the underlying file on
disk. `file_byte_offset()` for an open `segment_appender` will report
the current offset for the cursor, but since we `fallocate()` the file
up front, the actual file size is represented by `_fallocation_offset`.

Of course, the size reported by `file_byte_offset()` and `size_bytes()` will
be identical for a closed `segment_appender`. But for querying the size of
a segment with an open `appender`, `size_bytes()` should be used instead.

WillemKauf committed:
`storage`: use `_appender->size_bytes()` instead of `file_byte_offset()`

As mentioned, `file_byte_offset()` will not report the actual size of a file
on disk unless an `appender` is closed.

Use `_appender->size_bytes()` instead for proper reporting of the size of the
`.compaction_index` file on disk even when the `appender` is open.

dotnwat commented on src/v/storage/segment_appender.h:
```diff
@@ -84,6 +84,10 @@ class segment_appender {
         return _committed_offset + _bytes_flush_pending;
```
> But for querying the size of
a segment with an open `appender`, `size_bytes()` should be used instead.

would probably be a nice comment to add to file_byte_offset

dotnwat commented on src/v/storage/spill_key_index.cc:
```diff
@@ -401,7 +401,7 @@ std::ostream& operator<<(std::ostream& o, const spill_key_index& k) {
 
 size_t spill_key_index::size_bytes() const {
```
presumably this is only used for accounting, rather than any sort of decisions such as identifying if compaction occurred or not?

WillemKauf replied:
Correct

dotnwat commented on src/v/storage/segment.cc:
```diff
@@ -120,7 +120,9 @@ ss::future<usage> segment::persistent_size() {
      * using fallocate, always stat the on disk size for the head partition.
      */
-    if (!_appender && _data_disk_usage_size.has_value()) {
+    if (_appender) {
+        u.data = _appender->size_bytes();
+    } else if (_data_disk_usage_size.has_value()) {
         u.data = _data_disk_usage_size.value();
```
nice change

dotnwat commented on src/v/storage/segment.cc:
```diff
@@ -120,7 +120,9 @@ ss::future<usage> segment::persistent_size() {
      * segment appender will transparently extend the size of the segment file
      * using fallocate, always stat the on disk size for the head partition.
```
this comment needs updating

dotnwat reviewed (approved):
lgtm. i think maybe just a couple adjustments to comments is all that's needed.

feel free to put them into a different PR if it's convenient.
